Add support for running tests as inline PTX using NVRTC

ftvkyo opened the pull request:
Summary of changes:

- Support for dynamically loading the NVRTC library (the same way loading the CUDA library was implemented)
- Some refactoring to allow the PTX source files be inserted as inline PTX:
  - Test bodies use predictable argument names
  - Tests implement functions that return their PTX header, list of names of their arguments and their bodies separately
  - PTX to load the arguments is generated because it is different between PTX source (`ld.param`) and inline PTX (`mov`)
- `trait TestContext` that prepares the test body to be compiled + provides access to the CUDA library
  - implemented for `struct TestFixture<L>` where `L` is either `(Cuda,)` or `(Cuda, Nvrtc)` 
- When a test fails, binary, hexadecimal and decimal representation of mismatched values is printed to make debugging easier in certain cases

What do you think?

vosen reviewed (changes requested):
Good changes overall. Other than inline comments, some ptx tests started to fail:
```
cvt_f32_f16: FAIL with input 0.000000059604644775390625
    computed on GPU: 0.000000000000000000000000
    computed on CPU: 0.000000059604644775390625
cvt_f32_f32: OK
cvt_f64_f16: FAIL with input 0.000000059604644775390625
    computed on GPU: 0.000000000000000000000000
    computed on CPU: 0.000000059604644775390625
cvt_f64_f32: FAIL with input 0.000000000000000000000000
    computed on GPU: 0.000000000000000000000000
    computed on CPU: 0.000000000000000000000000
cvt_ftz_f32_f16: FAIL with input 0.000000059604644775390625
    computed on GPU: 0.000000000000000000000000
    computed on CPU: 0.000000059604644775390625
cvt_ftz_f32_f32: OK
cvt_ftz_f64_f32: FAIL with input 0.000000000000000000000000
    computed on GPU: 0.000000000000000000000000
    computed on CPU: 0.000000000000000000000000
cvt_ftz_sat_f32_f16: FAIL with input 0.000000059604644775390625
    computed on GPU: 0.000000000000000000000000
    computed on CPU: 0.000000059604644775390625
```
and `cvt_rm_f16_f32` crashes with
```
~/dev/ptx_pr1$ RUST_BACKTRACE=full cargo run  -- -f cvt_rm_f16_f32 /usr/lib/x86_64-linux-gnu/libcuda.so
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.04s
     Running `target/debug/ptx_tests -f cvt_rm_f16_f32 /usr/lib/x86_64-linux-gnu/libcuda.so`
thread 'main' panicked at src/test.rs:531:21:
called `Result::unwrap()` on an `Err` value: 218
stack backtrace:
   0:     0x5582028b884a - std::backtrace_rs::backtrace::libunwind::trace::h18d0110d4987d9e2
                               at /rustc/45d11e51bb66c2deb63a006fe3953c4b6fbc50c2/library/std/src/../../backtrace/src/backtrace/libunwind.rs:116:5
   1:     0x5582028b884a - std::backtrace_rs::backtrace::trace_unsynchronized::h145d10c8c28a6c5a
                               at /rustc/45d11e51bb66c2deb63a006fe3953c4b6fbc50c2/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x5582028b884a - std::sys::backtrace::_print_fmt::hfe84296be99dc012
                               at /rustc/45d11e51bb66c2deb63a006fe3953c4b6fbc50c2/library/std/src/sys/backtrace.rs:66:9
   3:     0x5582028b884a - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h4f6a4ee866651707
                               at /rustc/45d11e51bb66c2deb63a006fe3953c4b6fbc50c2/library/std/src/sys/backtrace.rs:39:26
   4:     0x5582028daa43 - core::fmt::rt::Argument::fmt::h7d9aaa4abd57f75c
                               at /rustc/45d11e51bb66c2deb63a006fe3953c4b6fbc50c2/library/core/src/fmt/rt.rs:177:76
   5:     0x5582028daa43 - core::fmt::write::h9e6f8d90343a57f6
                               at /rustc/45d11e51bb66c2deb63a006fe3953c4b6fbc50c2/library/core/src/fmt/mod.rs:1440:21
   6:     0x5582028b5ed3 - std::io::Write::write_fmt::h66c66b1419a58ace
                               at /rustc/45d11e51bb66c2deb63a006fe3953c4b6fbc50c2/library/std/src/io/mod.rs:1887:15
   7:     0x5582028b8692 - std::sys::backtrace::BacktraceLock::print::h498d080cb25470cd
                               at /rustc/45d11e51bb66c2deb63a006fe3953c4b6fbc50c2/library/std/src/sys/backtrace.rs:42:9
   8:     0x5582028b964d - std::panicking::default_hook::{{closure}}::hbbfe0167a7d38133
                               at /rustc/45d11e51bb66c2deb63a006fe3953c4b6fbc50c2/library/std/src/panicking.rs:279:22
   9:     0x5582028b9492 - std::panicking::default_hook::h9a3de122a21c6cbb
                               at /rustc/45d11e51bb66c2deb63a006fe3953c4b6fbc50c2/library/std/src/panicking.rs:306:9
  10:     0x5582028b9c42 - std::panicking::rust_panic_with_hook::hce10802d9af4ae90
                               at /rustc/45d11e51bb66c2deb63a006fe3953c4b6fbc50c2/library/std/src/panicking.rs:812:13
  11:     0x5582028b9ada - std::panicking::begin_panic_handler::{{closure}}::h970170fe06df4228
                               at /rustc/45d11e51bb66c2deb63a006fe3953c4b6fbc50c2/library/std/src/panicking.rs:685:13
  12:     0x5582028b8d19 - std::sys::backtrace::__rust_end_short_backtrace::hc705fc339aff58b7
                               at /rustc/45d11e51bb66c2deb63a006fe3953c4b6fbc50c2/library/std/src/sys/backtrace.rs:168:18
  13:     0x5582028b976d - rust_begin_unwind
                               at /rustc/45d11e51bb66c2deb63a006fe3953c4b6fbc50c2/library/std/src/panicking.rs:676:5
  14:     0x5582028d8370 - core::panicking::panic_fmt::h9cbd0b9e37a0849d
                               at /rustc/45d11e51bb66c2deb63a006fe3953c4b6fbc50c2/library/core/src/panicking.rs:75:14
  15:     0x5582028d87e6 - core::result::unwrap_failed::h507902d6dd0f7c52
                               at /rustc/45d11e51bb66c2deb63a006fe3953c4b6fbc50c2/library/core/src/result.rs:1704:5
  16:     0x5582024c3ad2 - core::result::Result<T,E>::unwrap::h784d7d97193ab93f
                               at /home/andrzej/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs:1109:23
  17:     0x5582024c3ad2 - ptx_tests::test::run_range::h7c7c2d1aaa13de36
                               at /home/andrzej/dev/ptx_pr1/src/test.rs:531:9
  18:     0x558202519fe8 - ptx_tests::test::make_range::{{closure}}::ha1bdd05b264f6a41
                               at /home/andrzej/dev/ptx_pr1/src/test.rs:648:32
  19:     0x5582025662f7 - core::ops::function::FnOnce::call_once{{vtable.shim}}::h0755b2544ed778ef
                               at /home/andrzej/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5
  20:     0x55820251b4b3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h47dd7ee3dc8d275f
                               at /home/andrzej/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:1990:9
  21:     0x55820251ae80 - ptx_tests::test::TestCase::new::{{closure}}::h8c9186e45e062fd0
                               at /home/andrzej/dev/ptx_pr1/src/test.rs:659:64
  22:     0x55820256637a - core::ops::function::FnOnce::call_once{{vtable.shim}}::h107543be0e9dff62
                               at /home/andrzej/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5
  23:     0x55820251b553 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h7317dcd4f05a9172
                               at /home/andrzej/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:1990:9
  24:     0x5582023056a6 - ptx_tests::run::h76e05ed539208077
                               at /home/andrzej/dev/ptx_pr1/src/main.rs:88:15
  25:     0x55820257e191 - ptx_tests::main::h779b70d9812205b6
                               at /home/andrzej/dev/ptx_pr1/src/main.rs:70:17
  26:     0x558202567d0b - core::ops::function::FnOnce::call_once::ha701fe6e3cf0b087
                               at /home/andrzej/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5
  27:     0x55820251b5be - std::sys::backtrace::__rust_begin_short_backtrace::h7f73dc46b8b53010
                               at /home/andrzej/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys/backtrace.rs:152:18
  28:     0x558202302251 - std::rt::lang_start::{{closure}}::h5018038cce05bd74
                               at /home/andrzej/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:195:18
  29:     0x5582028b2fa7 - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::h38eca8706ee4baee
                               at /rustc/45d11e51bb66c2deb63a006fe3953c4b6fbc50c2/library/core/src/ops/function.rs:284:13
  30:     0x5582028b2fa7 - std::panicking::try::do_call::hc5d26c6c9b591e0f
                               at /rustc/45d11e51bb66c2deb63a006fe3953c4b6fbc50c2/library/std/src/panicking.rs:568:40
  31:     0x5582028b2fa7 - std::panicking::try::h60e14511e9a61922
                               at /rustc/45d11e51bb66c2deb63a006fe3953c4b6fbc50c2/library/std/src/panicking.rs:531:19
  32:     0x5582028b2fa7 - std::panic::catch_unwind::h9f5e2231130cc217
                               at /rustc/45d11e51bb66c2deb63a006fe3953c4b6fbc50c2/library/std/src/panic.rs:358:14
  33:     0x5582028b2fa7 - std::rt::lang_start_internal::{{closure}}::h90199df0ee7f84bc
                               at /rustc/45d11e51bb66c2deb63a006fe3953c4b6fbc50c2/library/std/src/rt.rs:174:48
  34:     0x5582028b2fa7 - std::panicking::try::do_call::h2c083b889f6a6279
                               at /rustc/45d11e51bb66c2deb63a006fe3953c4b6fbc50c2/library/std/src/panicking.rs:568:40
  35:     0x5582028b2fa7 - std::panicking::try::hf476aebcb5943b3e
                               at /rustc/45d11e51bb66c2deb63a006fe3953c4b6fbc50c2/library/std/src/panicking.rs:531:19
  36:     0x5582028b2fa7 - std::panic::catch_unwind::h1e796d75a8c44324
                               at /rustc/45d11e51bb66c2deb63a006fe3953c4b6fbc50c2/library/std/src/panic.rs:358:14
  37:     0x5582028b2fa7 - std::rt::lang_start_internal::hd53cb9d9cbc5a6b9
                               at /rustc/45d11e51bb66c2deb63a006fe3953c4b6fbc50c2/library/std/src/rt.rs:174:20
  38:     0x55820230222a - std::rt::lang_start::h9fa488e3b453af6b
                               at /home/andrzej/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:194:17
  39:     0x55820257ed9e - main
  40:     0x7fdda52bdd90 - __libc_start_call_main
                               at ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
  41:     0x7fdda52bde40 - __libc_start_main_impl
                               at ./csu/../csu/libc-start.c:392:3
  42:     0x5582022f58a5 - _start
  43:                0x0 - <unknown>
```

I did not check nvrtc tests, because I assume you have your eyes on them

ftvkyo commented:
Thank you for the directions -- I read through them and things seem to be clear.
I will get to implementing the changes next week.

Just this week I noticed that there were issues with using CUDA's NVRTC that I missed as I thought I ran some tests, when in fact I didn't.
I will change my workflow to make sure I test it better.

On that note -- currently, the CPU is really the bottleneck when running these tests.
I have been thinking -- really, the test suite tests 2 different things:
1. Whether the PTX / inline PTX code builds successfully
2. Whether PTX instructions are implemented correctly

I think the former is more likely to break with changes to the test suite and/or ZLUDA's / SCALE's PTX parsing code.
It might be a good idea to add a "dry run" option to the test suite that builds all tests, but does not run them.
This is simple enough for this PR, so I want to add it -- it will make it much easier to make sure I didn't break existing things with this PR.

The other thing we thought about at Spectral Compute is whether it would be a good idea to generate and save test inputs and expected outputs, so that the CPU stops being so much of a bottleneck when running the tests.
This is definitely beyond the current scope, so it's just an idea for another time.

vosen commented:
My current thinking is that developer works on a single test at a time and whole suite runs overnight. If you work on a single test then you will notice parsing failure very quickly and from that you can minimize the ptx to repro just the part of ptx that causes failure and add to your project's ptx parser or compiler test suite. In nightly runs, right now you will get crashes with invalid ptx, but that's only in rng tests because I was too lazy to handle it gracefully - that's the place that could be improved. I'm not sure how dry run is helpful, what sort of workflow does it fit into?

On the topic of storing input and output: in the ideal world that would be the solution. The problem is storage. You need 16GB minimum for the output of an fp32 test (2^32 * 4 bytes) and Github caps file sizes at 2GB and total repo size at 100GB.

ftvkyo commented:
I see your point -- indeed, there is no reason for a dry run option if the tool is used the way you described.

Our current use case is a bit wider than that -- `ptx_tests` is useful not just for finding problems with how exactly PTX instructions are calculated, but also for making sure that SCALE's behaviour is compatible with the behaviour of `nvcc` and NVRTC.

Specifically, when implementing fixes to implementations of PTX instructions in SCALE, I found myself going through the following loop at times:
1. Build and install SCALE
2. Run `ptx_tests`
3. Possibly get a compilation problem or a compiler crash, fix the problem and return to 1

Sometimes the problem may present itself only with some of the tests and not the others (e.g. bfi.b32 vs bfi.b64), so to see if the thing builds, I basically need to run multiple instances of `ptx_tests` with the intention of stopping them as soon as it's clear that my changes didn't break PTX parsing and associated IR emission.

In these circumstances, `ptx_tests` essentially takes the place of unit tests that a developer runs regularly, and that's why I have been considering ways to make that loop easier to go through.
I also work on SCALE's unit tests as I go along though, so this is more of a one-off thing for each instruction.
At the same time, I imagine eventually there will be more instructions that we will want to test this way.
In any case, it's totally fine if dry run is unnecessary here, I can easily keep it in a separate branch in my repo.

vosen commented:
And I forgot on the topic of performance: I did not look into the assembly, but I suspect the current host-side function compute and check can't be vectorized. If the verification loop were to be done over a slice instead of one-by-one it'd probably be a candidate for auto-vectorization (unless the test is using `rug`)

ftvkyo commented on src/test.rs:
```diff
@@ -1,15 +1,28 @@
+use crate::TestContext;
+
+pub trait TestPtx {
+    fn header(&self) -> &str {
```
Implemented, now the header is
```
    .version 7.0
    .target sm_80
    .address_size 64
```

And I also pass an `-arch=sm_80` option to NVRTC

ftvkyo commented on src/test.rs:
```diff
@@ -282,8 +295,45 @@ impl<X: OnDevice, Y: OnDevice, Z: OnDevice, W: OnDevice> OnDevice for (X, Y, Z,
 }
 
-pub trait PtxScalar: Copy + Num + Bounded + Debug + OnDevice + Any {
+pub trait DebugRich {
```
Implemented for all inputs, outputs & expected.
Also implemented for tuples of `DebugRich` up to the length of 4.

ftvkyo commented on src/test.rs:
```diff
@@ -282,8 +295,45 @@ impl<X: OnDevice, Y: OnDevice, Z: OnDevice, W: OnDevice> OnDevice for (X, Y, Z,
+impl_debug_rich!(i32);
+impl_debug_rich!(u64);
+impl_debug_rich!(i64);
+impl_debug_rich!(f16);
```
That was done this way because f16 comes from a library, and it implements `std::fmt::UpperHex` and `std::fmt::Binary`, but as we want to specify precision, I implemented it separately as requested.

ftvkyo commented:
The CVT tests started failing because of a mistake I made implementing
```
impl<To: PtxScalar, From: PtxScalar> TestPtx for Cvt<To, From> {
    fn body(&self) -> String {
        /* ... */
    }
}
```

I wrote

```
        let input_bits = mem::size_of::<From>() * 8;
        let output_bits = mem::size_of::<From>() * 8;
```

Instead of

```
        let input_bits = mem::size_of::<From>() * 8;
        let output_bits = mem::size_of::<To>() * 8;
```

This is now fixed.

I didn't find other mistakes like this, but I am running the full test suite now to make sure there is nothing else.
It takes a while though.

Additionally, in a personal branch, I implemented the dry run I mentioned: https://github.com/vosen/ptx_tests/commit/40d369dbaf1e493f38e078af832e1cf42d7f0cc7
I ran it against the CUDA lib, and CUDA + NVRTC, so while I don't yet have the results of a complete run, I can say that modules load and produced inline PTX compiles.

ftvkyo commented:
The following tests fail, I will look into them next week:

- lg2_approx
- max
- max_ftz
- max_ftz_nan
- max_nan
- min
- min_ftz
- min_ftz_nan
- min_nan

vosen commented on src/main.rs:
```diff
@@ -45,67 +44,76 @@ pub enum Arguments {
+
+            let failures = if let Some(nvrtc) = nvrtc {
+                let libs = (cuda, nvrtc);
+                run(tests, TestFixture { libs })
```
Ahh, I did not realize

vosen commented:
Thanks, nice changes 👍
